mypy/typestate.py

```diff
@@ -0,0 +1,55 @@
+"""
+A shared state for all TypeInfo's that holds global cache and dependency information,
+and potentially other mutable TypeInfo state.
+"""
+
+from typing import Dict, Set, Tuple
+
+MYPY = False
+if MYPY:
+    from typing import ClassVar
+    from mypy.nodes import TypeInfo
+    from mypy.types import Instance
+
+
+class TypeState:
+    # 'caches' and 'caches_proper' are subtype caches, implemented as sets of pairs
+    # of (subtype, supertype), where supertypes are instances of given TypeInfo.
+    # We need the caches, since subtype checks for structural types are very slow.
+    caches = {}  # type: ClassVar[Dict[TypeInfo, Set[Tuple[Instance, Instance]]]]
+    caches_proper = {}  # type: ClassVar[Dict[TypeInfo, Set[Tuple[Instance, Instance]]]]
+
+    @classmethod
+    def reset_all(cls) -> None:
+        cls.caches = {}
+        cls.caches_proper = {}
+
+    @classmethod
+    def add_caches(cls, info: 'TypeInfo') -> None:
```

Comment:
As discussed above, this seems unnecessary in the public API. We can create the caches on demand here.